pkg/rehearse/jobs.go

```diff
@@ -203,25 +209,114 @@ func ConfigureRehearsalJobs(toBeRehearsed config.Presubmits, ciopConfigs config.
 				continue
 			}
 
-			exists, index, templateKey := hasChangedTemplateVolume(rehearsal.Spec.Containers[0].VolumeMounts, rehearsal.Spec.Volumes, templates)
+			exists, index, cmName, err := getVolumeIndexAndTempCMName(rehearsal.Spec.Containers[0].VolumeMounts, rehearsal.Spec.Volumes, templates)
+			if err != nil {
+				loggers.Job.WithError(err).Warn(skipJobMessage)
+				continue
+			}
+
 			if exists {
-				if templateData, err := config.GetTemplateData(templates[templateKey]); err != nil {
-					jobLogger.WithError(err).WithField("template-name", templates[templateKey].Name).Warn("couldn't get template's data. Job won't be rehearsed")
-					continue
-				} else {
-					templateName := config.GetTemplateName(templateKey)
-					rehearsal.Spec.Volumes[index].VolumeSource.ConfigMap.Name = config.GetTempCMName(templateName, templateKey, templateData)
-				}
+				rehearsal.Spec.Volumes[index].VolumeSource.ConfigMap.Name = cmName
 			}
 
 			jobLogger.WithField(logRehearsalJob, rehearsal.Name).Info("Created a rehearsal job to be submitted")
 			rehearsals = append(rehearsals, rehearsal)
 		}
 	}
 
+	// for the remaining cluster types of each template, pick a `random` corresponding job to rehearse.
+	for templateFile, clusterTypes := range templatesWithClusterTypes {
+		for clusterType := range clusterTypes {
+			if repo, job := pickTemplateJob(masterConfigPresubmits, templateFile, clusterType); job != nil {
+				jobLogger := loggers.Job.WithFields(logrus.Fields{"target-repo": repo, "target-job": job.Name})
+				rehearsal, err := makeRehearsalPresubmit(job, repo, prNumber)
+				if err != nil {
+					jobLogger.WithError(err).Warn("Failed to make a rehearsal presubmit")
+					continue
+				}
+				exists, index, cmName, err := getVolumeIndexAndTempCMName(job.Spec.Containers[0].VolumeMounts, job.Spec.Volumes, templates)
+				if err != nil {
+					loggers.Job.WithError(err).Warn(skipJobMessage)
+					continue
+				}
+
+				if exists {
+					rehearsal.Spec.Volumes[index].VolumeSource.ConfigMap.Name = cmName
+				}
+
+				rehearsals = append(rehearsals, rehearsal)
+			}
+		}
+	}
+
 	return rehearsals
 }
 
+func getVolumeIndexAndTempCMName(volumeMounts []v1.VolumeMount, volumes []v1.Volume, templates config.CiTemplates) (bool, int, string, error) {
```

Comment:
Returning four items is pushing readability limits. Maybe we could do:

```
type VolumeRef struct {
  index int
  name string
}
```
and let the function return `(*VolumeRef, error)` where `nil` means not found.

Just an idea, not a must address.

Reply:
I was thinking that too, but I didn't do it, because it's being used only 2 times for the templates.

Reply:
addressed!